rust/arrow/src/compute/kernels/boolean.rs

```diff
@@ -457,4 +517,42 @@ mod tests {
         assert_eq!(true, res.value(2));
         assert_eq!(false, res.value(3));
     }
+
+    fn assert_array_eq<T: ArrowNumericType>(
+        expected: PrimitiveArray<T>,
+        actual: ArrayRef,
+    ) {
+        let actual = actual
+            .as_any()
+            .downcast_ref::<PrimitiveArray<T>>()
+            .expect("Actual array should unwrap to type of expected array");
+
+        for i in 0..expected.len() {
+            if expected.is_null(i) {
+                assert!(actual.is_null(i));
+            } else {
+                assert_eq!(expected.value(i), actual.value(i));
+            }
+        }
+    }
+
```

Comment:
I wrote the following test with arrays with an offset, as they are typical sources of errors, but I was unable to make it pass (even after fixing the null count):

```rust
    #[test]
    fn test_nullif_int_array_offset() {
        let a = Int32Array::from(vec![None, Some(15), Some(8), Some(1), Some(9)]);
        let a = a.slice(1, 3); // Some(15), Some(8), Some(1)
        let comp =
            BooleanArray::from(vec![Some(false), Some(false), Some(false), None, Some(true), Some(false), None]);
        let comp = comp.slice(2, 3); // Some(false), None, Some(true)
        let comp = comp.as_any().downcast_ref::<BooleanArray>().unwrap();
        let res = nullif(a.as_ref(), comp).unwrap();

        let expected = Arc::new(Int32Array::from(vec![
            Some(15), // False => keep it
            Some(8), // None => keep it
            None, // true => None
        ])) as ArrayRef;
        assert_eq!(&expected, &res)
    }
```

Do you know what happens when the arrays have offsets?

Reply:
It's a good question @jorgecarleitao and thanks so much for digging into it.    I had found some mention of offsets but haven't completely grokked what was going on there, looks like will have to dive into it a little bit.

I'll rebase and look into the data() discrepancy above and the details of the array equality comparison, and report back.  cheers and thanks!

Reply:
An offset is just a shift of where we consider reading a buffer. The easiest way to motivate it: suppose that we want to slice a primitive array of 100 elements over the range `[25..75]`. The expensive way of doing this is a `memcopy` of 50 elements. The cheap way is to declare a new array with len=50 elements starting at 25 and increase the refcount of buffer (`Arc::clone` in Rust).

The "starting at 25" is an offset of 25 "slots", or `25 * size_of::<T>()` bytes for a specific type T.

For bit buffers (nullability), the offset is measured in bits, not bytes. Feel free to ping e.g. @vertexclique, @alamb or @jhorstmann , that lately have been working (a lot) on bits - they likely know exactly the API you need for this. :)

Reply:
I think I figured it out....   though an example would be great.  Basically, to do bitmap/buffer ops on offsets, you have to use like `bit_slice()` to get a normalized map (based on 0), then shift it back.

Reply:
I don't see an API to shift bits forward, which I think would need to make this work....

Reply:
To give more details on the issue:  we have basically a left_bitmap and a right_bitmap.  The left_bitmap is at left_offset_bits bits, and the right bitmap is at right_offset_bits.

If left_offset_bits == right_offset_bits, then everything is easy, the resulting bitmap in fact does not need to be shifted.

If left_offset_bits != right_offset_bits, one can do:
* Use bit_slice() on left_bitmap and on right_bitmap, the result is 0-based and can be operated on
* After the operation, "right shift" the resulting bitmap so it is based on left_offset again (since the resulting data is based on left array)

An optimization of the above is to shift right_bitmap so it lines up with left_bitmap.

I see `bit_slice()`, but I don't see anything in the Buffer API to "right shift".....

@vertexclique @alamb any ideas?  Thanks.

Reply:
In my PR if you use `slicing` method and then write a method to align bit slices with smth along the lines:
```
        unsafe {
            let (_, data, _) = self.bit_slice.align_to::<u8>();
            let dest = data.as_slice();
         }
```
you will get two aligned middle bit domains. then you can check bit slices however you want to.

slicing method:
https://github.com/apache/arrow/pull/8664/files#diff-715ddfbc534281523a73117d3bf4986d69d8a375dd320bc19f83298d3ba7ebd6R52

Reply:
@velvia, take a look at `bitwise_unary_op_helper` on `src/buffer.rs`. I think that inside it, there is functionality relevant to this.

Reply:
@vertexclique ... I had a quick look at the PR but slicing() method was not defined in it.  Is it from bitvec crate or something?

@jorgecarleitao well `bitwise_unary_op_helper` (and `bitwise_bin_op_helper` also) both use `bit_chunks()` which basically gives you an iterator of 64-bit chunks starting at offset.... essentially "left-shifting" into 0 position.

I need a reverse operation, which is take 64-bit chunks and right shift them.  It can be kind of simulated by "pre-pending" 64 0-bits into a buffer, then using bit_chunks() again to get a shift into whatever alignment is needed, and finally writing out the final buffer, though all of that is slightly convoluted.

I'm wondering if it's worth trying that out, or just wait for @vertexclique 's PR.

Actually, what's better than shift-shift-zip-operate-rshift would be simply to shift the right bitmap into the left's position, then use a mask to operate on both of them.  That reduces the number of shifts.

Reply:
@velvia In the current version you could use the `bitwise_bin_op_helper`from `buffer.rs` if you make that public. That calls your bitwise operation with two `u64` as an input and takes care of the offset alignment.

In our internal codebase we have similar functions also for 3 or 4 inputs, which would come in handy for you usecase, since you could calculate the result without the temporary `right_combo_buffer`. So you could also think about copying that function and modifying to take a function of 3 parameters instead.

Reply:
@jhorstmann  Ok let me try to explain.

The nullif function takes the original array, which let's say has offset=2.   It uses the result of another boolean array, which let's say has offset=3.   The nullif kernel must combine both the original nullif null bitmap at offset=2 with the boolean array values and null bitmaps (both at offset=3), and after combining, produce a new array with a combined bitmap at offset=2 (so we can produce a new array, with the same data as the original left array, but a modified nullity bitmap).

Using `bitwise_bin_op_helper` (actually I can just use `bit_slice()` on original bitmaps, though it's a bit slower) takes care of the first part - basically it takes two bitmaps of different alignments, aligns them and does bit operations on 64-bit aligned words.  What I end up with is combined(left_bitmap, right_bitmap) aligned at offset 0.

What is missing to construct the final nullif is the above combined bitmap but at offset 2, so it can be aligned with the original left data.   I need to insert two bits in the LSB of the resulting bitmap so it is aligned with the original data.   AFAICT there isn't anything that can do this in the current code.

Reply:
@velvia Thanks, I got it now. It was already quite late when I looked at this yesterday :)

Reusing the input array including its offset is a tricky problem, it's also not always beneficial from a performance standpoint, the offset might start at a much larger number and you probably do not want to create a null buffer with that much unused space in front. I see two possible solutions then:

- Use the `buffer.slice(offset_bytes)` method for also getting a data slice that can then be used with an offset of 0. Requires a special case for primitive type boolean, for which you'd need to use `bit_slice(offset_in_bits)`.

- Create a copy of the input data if the offset is not 0.

Reply:
You could probably avoid the special case in the first option by always using `bit_slice` on the data buffers, but calculating the offset based on `ArrowPrimitiveType::get_bit_width`. That and using `bit_slice` for the boolean and null bitmaps looks like the cleanest solution to me.

Reply:
> What is missing to construct the final nullif is the above combined bitmap but at offset 2, so it can be aligned with the original left data. I need to insert two bits in the LSB of the resulting bitmap so it is aligned with the original data. AFAICT there isn't anything that can do this in the current code.

That's why we need to have bit-slice iterator views on the code, rather than copying slices like suggested before. Bit domain view exposed in the PR that I created can give you what you need by force aligning the msb. You can reuse the code everywhere, without `actually` copying anything.

That was one of the ideas why I introduced that PR here.

Reply:
Thanks everyone for their comments.  

It seems to me that taking a slice of the data buffer would be even more expensive than shifting the null buffer?  However it is easier to implement for now, so I will try @jhorstmann 's suggestion using `bit_slice()` on the data buffers - would rather get in something that works sooner and iterate later when @vertexclique 's PR is merged.